crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -171,82 +213,767 @@ impl QubitTracker {
     }
 }
 
-/// Correspondence between local qubits and global qubits.
-///
-/// An internal class for handling recursion within `HighLevelSynthesis`.
-/// Provides correspondence between the qubit indices of an internal DAG,
-/// aka the "local qubits" (for instance, of the definition circuit
-/// of a custom gate), and the qubit indices of the original DAG, aka the
-/// "global qubits".
-///
-/// Since the local qubits are consecutive integers starting at zero,
-/// i.e. `0`, `1`, `2`, etc., the correspondence is kept using a vector, with
-/// the entry in position `k` representing the global qubit that corresponds
-/// to the local qubit `k`.
-#[pyclass]
+/// Internal class that encapsulates immutable data required by the HighLevelSynthesis transpiler pass.
+#[pyclass(module = "qiskit._accelerate.high_level_synthesis")]
 #[derive(Clone, Debug)]
-pub struct QubitContext {
-    /// Mapping from local indices to global indices
-    local_to_global: Vec<usize>,
+struct HighLevelSynthesisData {
+    // The high-level-synthesis config that specifies the synthesis methods
+    // to use for high-level-objects in the circuit.
+    // This is only accessed from the Python space.
+    #[pyo3(get)]
+    hls_config: Py<PyAny>,
+
+    // The high-level-synthesis plugin manager that specifies the synthesis methods
+    // available for various high-level-objects.
+    // This is only accessed from the Python space.
+    #[pyo3(get)]
+    hls_plugin_manager: Py<PyAny>,
+
+    // The names of high-level objects with available synthesis plugins.
+    // This is an optimization to avoid calling python when an object has no
+    // synthesis plugins.
+    #[pyo3(get)]
+    hls_op_names: Vec<String>,
```

Comment:
Would it make sense to use a HashSet here too, given that we're always checking whether some op name is in this vector?

Reply:
Done in [51e0662](https://github.com/Qiskit/qiskit/pull/13813/commits/51e0662c127f9b9f34858ea244660e8c302365c9).